src/sources/kafka.rs

```diff
@@ -374,105 +390,406 @@ impl SourceConfig for KafkaSourceConfig {
+    }
+
+    pub fn clear(&mut self) {
+        self.expect_drain.clear();
+        self.observed_drain.clear();
+    }
+}
+
+async fn coordinate_kafka_callbacks(
+    consumer: Arc<StreamConsumer<KafkaSourceContext>>,
+    mut callbacks: UnboundedReceiver<KafkaCallback>,
+    mut partition_state: KafkaPartitionState,
+    max_drain_ms: Duration,
+    mut eof: Option<tokio::sync::oneshot::Sender<()>>,
+) {
+    let mut drain_signal: Option<SyncSender<()>> = None;
+
+    async fn revoke_timeout(t: Duration) {
+        tokio::time::sleep(t).await;
     }
+    let mut drain_deadline = tokio::spawn(revoke_timeout(max_drain_ms));
+
+    // A oneshot channel is used for each consumed partition, so that we can
+    // signal to that task to stop consuming, drain pending acks, and exit
+    let mut end_signals: HashMap<TopicPartition, tokio::sync::oneshot::Sender<()>> = HashMap::new();
 
-    let mut stream = consumer.stream();
+    // The set of consumer tasks, each consuming a specific partition. The task
+    // is both consuming the messages (passing them to the output stream) _and_
+    // processing the corresponding acknowledgement stream. A consumer task
+    // should completely drain its acknowledgement stream after receiving an end signal
+    let mut partition_consumers: tokio::task::JoinSet<(TopicPartition, PartitionConsumerStatus)> = Default::default();
+
+    // Handles that will let us end any consumer task that exceeds a drain deadline
+    let mut abort_handles: HashMap<TopicPartition, tokio::task::AbortHandle> = HashMap::new();
+
+    let exit_eof = eof.is_some();
 
     loop {
         tokio::select! {
-            biased;
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
-                    }
+            Some(Ok((finished_partition, status))) = partition_consumers.join_next(), if !partition_consumers.is_empty() => {
+                debug!("Partition consumer finished for {}:{}", &finished_partition.0, finished_partition.1);
+                abort_handles.remove(&finished_partition);
+                partition_state.observed_last_ack(finished_partition);
+
+                // Signal the client task that at least one partition has completed
+                _ = drain_signal.as_ref().map(|sig| _ = sig.send(()) );
+
+                if partition_state.is_drain_complete() {
+                    debug!("All expected partitions have drained.");
+                    partition_state.clear();
+                    // Signal the client task that all partitions that are expected to drain have completed
+                    drain_signal.take();
+                }
```

Comment:
In conjunction with the suggestion for `is_drain_complete`, we should add a check here to verify that there is an active drain in progress.
```suggestion
                if drain_signal.is_some() && partition_state.is_drain_complete() {
                    debug!("All expected partitions have drained.");
                    partition_state.clear();
                    // Signal the client task that all partitions that are expected to drain have completed
                    drain_signal.take();
                }
```